### PR TITLE
Script for collecting task files paths.

### PR DESCRIPTION
**Pull request by mikrise2**

### CODE REVIEW

File: jba/README.md [hunk 1]
```diff
@@ -84,6 +84,25 @@ docker run hyperstyle-analysis-prod:<VERSION> poetry run process_course_data [ar
...
+### Usage
+
+Execute one of the following commands with necessary arguments:
+```bash
+poetry run tasktracker_content_collector [arguments]
+```
+or
+```bash
+docker run hyperstyle-analysis-prod:<VERSION> poetry run tasktracker_content_collector [arguments]
+```
```
Comment (GirZ0n) [inline jba/README.md:96]: The name of the run command should be the same as the name defined in the `pyproject.toml` file

File: jba/README.md [hunk 2]
```diff
@@ -84,6 +84,25 @@ docker run hyperstyle-analysis-prod:<VERSION> poetry run process_course_data [ar
 
 After this step you will get a new file with course data with `courseId_preprocessed` suffix. 
 
+
+3. [tasktracker_content_collector.py](src/processing/tasktracker_content_collector.py) Collects data from the course into the tasktracker task content file format
```
Comment (GirZ0n) [inline jba/README.md]: **c**ollects

File: jba/src/processing/tasktracker_content_collector.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    def __eq__(self, other):
+        if isinstance(other, TaskTrackerFile):
+            return self.name == other.name and self.path == other.path and self.extension == other.extension
+        return False
+
+    def __hash__(self):
+        return hash((self.name, self.path, self.extension))
+
+
+def get_data_template(files: list) -> dict:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Please use typings from the `typing` library

File: jba/src/processing/tasktracker_content_collector.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    parser.add_argument(
+        'course_sources_path',
+        type=lambda value: Path(value).absolute(),
+        help='Path to course sources to extract course structure.',
+    )
+
+
+def save_yaml_file(yaml_content: dict, path: Path):
+    with open(path, 'w') as file:
+        yaml.dump(yaml_content, file)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Move it to the `yaml_utils.py` file and rename to `save_as_yaml` please

File: jba/src/processing/tasktracker_content_collector.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+TYPE_META_FIELD = 'type'
+NAME_META_FIELD = 'name'
+
+CONTENT_FILE_NAME = 'task_content_default.yaml'
+
+FRAMEWORK_TYPE = 'framework'
+
+TASK_DIRECTORY_NAME = 'task'
+
+INFO_FILE_REGEX = re.compile(f'([a-z]+)-info{AnalysisExtension.YAML.value}')
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Use the regex from the `prepare_course_data.py` file

File: jba/src/processing/tasktracker_content_collector.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def get_info_file(root: Path) -> str:
+    file_names = listdir(root)
+    info_files = list(filter(lambda file_name: re.match(INFO_FILE_REGEX, file_name), file_names))
+
+    return info_files[0]
+
+
+def get_lessons(root: Path):
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: If think we should gather the course structure using the `_gather_structure` function in the `prepare_course_data.py` file (probably it should be renamed to something like `gather_structure` and the `get_course_structure` function should also be renamed to `convert_structure_to_dataframe` and it should accept `course_structure: EduStructureNode`) and then using tree traversal find all task files. 

Because right now this looks like `prepare_course_data.py` but with some additional logic. And I think that the traversal can be done either inside the `EduStructureNode` class or inside the `edu_structure.py` file and maybe the traverse function should be able to return not only tasks, but any items with the EduStructureType (it should be controlled by the function argument).

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    result = {}
+    with open(my_file, 'r') as file_path:
+        docs = yaml.safe_load_all(file_path)
+        for doc in docs:
+            for key, value in doc.items():
+                result[key] = value
+    return result
+
+
+def test_correct_arguments():
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: Add more test cases, please

**Review by GirZ0n — CHANGES_REQUESTED** [review]
I left several comments. If you have any questions, message me in Slack :-)

**Comment by GirZ0n** [comment]
Nastya've merged my PR, so the `prepare_course_data.py` file is now called `collect_course_structure.py`

File: jba/src/processing/tasktracker_content_collector.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    def get_filename(file_content: Dict) -> TaskTrackerFile:
+        if is_framework:
+            return TaskTrackerFile(relative_path / TASK_DIRECTORY_NAME / file_content[NAME_META_FIELD])
+        return TaskTrackerFile(relative_path / root.name / file_content[NAME_META_FIELD])
+
+    return list(map(get_filename, files))
+
+
+def course_structure_to_lessons(root: Path, structure: EduStructureNode):
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Use the `EduStructureNode.gather_leafs_of_type` function instead of this one

File: jba/src/processing/tasktracker_content_collector.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    file_names = listdir(root)
+    info_files = list(filter(lambda file_name: re.match(INFO_FILE_REGEX, file_name), file_names))
+
+    return info_files[0]
+
+
+def get_files(root: Path, lesson: EduLesson) -> List[TaskTrackerFile]:
+    relative_path = lesson.root.relative_to(root)
+    return flatten(
+        [get_task_files(lesson.root / child, relative_path, lesson.is_framework) for child in lesson.children])
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `get_task_files` always returns a list of simple (unnested) lists so you can use `itertools.chain.from_iterable` here

File: jba/src/processing/tasktracker_content_collector.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+def flatten(files: List) -> List[TaskTrackerFile]:
+    result = []
+    for i in files:
+        if i is None:
+            continue
+        if isinstance(i, list):
+            result.extend(flatten(i))
+        else:
+            result.append(i)
+    return result
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: And it seems you can delete this function

File: jba/src/processing/tasktracker_content_collector.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+
+def get_info_file(root: Path) -> str:
+    file_names = listdir(root)
+    info_files = list(filter(lambda file_name: re.match(INFO_FILE_REGEX, file_name), file_names))
+
+    return info_files[0]
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Use `find_files_by_regex` from `file_utils.py`

File: jba/src/models/edu_structure.py [hunk 12]
```diff
@@ -16,3 +32,23 @@ class EduStructureNode:
...
+            for child in self.children
+            for path, leaf in child.gather_leafs_of_type(leaf_type).items()
+        }
+
+
+@dataclass(frozen=True)
+class EduLesson:
+    root: Path
+    is_framework: bool
+    children: List[str]
```
Comment (GirZ0n) [inline jba/src/models/edu_structure.py]: I think we can do without this dataclass, delete it and use `EduStructureNode`

File: jba/src/processing/tasktracker_content_collector.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+
+def get_yaml_content(course_root: Path) -> Dict:
+    course_structure = gather_structure(course_root)
+    lessons = course_structure_to_lessons(course_root, course_structure)
+
+    files = set()
+    for lesson in lessons:
+        for file in get_files(course_root, lesson):
+            files.add(file)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Use a set comprehension here

File: jba/src/processing/tasktracker_content_collector.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    return get_data_template(list(map(lambda obj: obj.as_dict(), files)))
+
+
+def configure_parser(parser: argparse.ArgumentParser) -> None:
+    parser.add_argument(
+        'destination_path',
+        type=lambda value: Path(value).absolute(),
+        help='Path to directory where yaml file will be created',
+    )
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Maybe we should pass the file path instead of the folder path?

File: jba/src/processing/tasktracker_content_collector.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    INFO_FILE_REGEX,
+)
+
+CONTENT_FILE_NAME = 'task_content_default.yaml'
+FRAMEWORK_TYPE = 'framework'
+TASK_DIRECTORY_NAME = 'task'
+EXTENSIONS = {'py': 'PYTHON', 'ipynb': 'JUPYTER', 'java': 'JAVA', 'kt': 'KOTLIN', 'cpp': 'CPP', 'csv': 'CSV'}
+
+
+@dataclass(eq=False, frozen=True)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Why did you disable `eq`?

File: jba/src/processing/tasktracker_content_collector.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    return info_files[0]
+
+
+def get_files(root: Path, lesson: EduLesson) -> List[TaskTrackerFile]:
+    relative_path = lesson.root.relative_to(root)
+    return flatten(
+        [get_task_files(lesson.root / child, relative_path, lesson.is_framework) for child in lesson.children])
+
+
+def get_task_files(root: Path, relative_path: Path, is_framework: bool):
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Add a return type

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        file = Path(temp_directory) / 'task_content_default.yaml'
+        assert file.exists()
+        expected_file = (PROCESSING_FOLDER / 'tasktracker_content_collector' / f'expected_{course_name}.yaml')
+        difference = DeepDiff(yaml_as_dict(file), yaml_as_dict(expected_file), ignore_order=True)
+        assert not difference
+
+
+def test_correct_arguments():
+    _test_course('course_with_section')
+    _test_course('course_without_section')
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: Why did you delete the third test case?

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        assert stderr == ''
+        assert len(os.listdir(temp_directory)) == 1
+        file = Path(temp_directory) / 'task_content_default.yaml'
+        assert file.exists()
+        expected_file = (PROCESSING_FOLDER / 'tasktracker_content_collector' / f'expected_{course_name}.yaml')
+        difference = DeepDiff(yaml_as_dict(file), yaml_as_dict(expected_file), ignore_order=True)
+        assert not difference
+
+
+def test_correct_arguments():
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: Please use parametrized tests. Look at `test_collect_course_structure.py`

Comment (GirZ0n) [reply jba/tests/processing/test_tasktracker_content_collector.py]: And add the test case when a course has a framework lesson

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def yaml_as_dict(my_file):
+    result = {}
+    with open(my_file, 'r') as file_path:
+        docs = yaml.safe_load_all(file_path)
+        for doc in docs:
+            for key, value in doc.items():
+                result[key] = value
+    return result
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: You can use `parse_yaml` from `yaml_utils.py`

**Review by GirZ0n — CHANGES_REQUESTED** [review]

File: jba/src/processing/tasktracker_content_collector.py (hunk 14 above)
Comment (mikrise2) [reply jba/src/processing/tasktracker_content_collector.py]: The point is that we don't want to let the user choose the configuration file name because configuration files are identified by names in the TT

File: jba/tests/processing/test_tasktracker_content_collector.py (hunk 17 above)
Comment (mikrise2) [reply jba/tests/processing/test_tasktracker_content_collector.py]: Because when using gather_structure, it throws an error, so maybe it's worth adding it to the list of tests where the structure is incorrect.

File: jba/tests/processing/test_tasktracker_content_collector.py (hunk 19 above)
Comment (mikrise2) [reply jba/tests/processing/test_tasktracker_content_collector.py]: I need to use yaml.safe_load_all while parse_yaml() uses yaml.safe_load

File: jba/src/processing/tasktracker_content_collector.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    INFO_FILE_REGEX,
+)
+
+CONTENT_FILE_NAME = 'task_content_default.yaml'
+FRAMEWORK_TYPE = 'framework'
+TASK_DIRECTORY_NAME = 'task'
+EXTENSIONS = {'py': 'PYTHON', 'ipynb': 'JUPYTER', 'java': 'JAVA', 'kt': 'KOTLIN', 'cpp': 'CPP', 'csv': 'CSV'}
+
+
+@dataclass(eq=True, frozen=True)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `eq` defaults to `True`, so `eq=True` can be removed

File: jba/src/processing/tasktracker_content_collector.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        }]
+    }
+
+
+def get_info_file(root: Path) -> Path:
+    info_files = find_files_by_regex(root, INFO_FILE_REGEX)
+    return info_files[0]
+
+
+def get_files(course_root: Path, relative_path: Path, lesson: EduStructureNode) -> List[TaskTrackerFile]:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `get_files` -> `get_lesson_files`

File: core/src/utils/file/file_utils.py [hunk 22]
```diff
@@ -94,3 +94,7 @@ def get_output_path(input_path: Union[str, Path], output_suffix: str) -> Path:
     output_filename = get_output_filename(input_path, output_suffix)
 
     return parent_dir / output_filename
+
+
+def find_files_by_regex(path: Path, regex: re.Pattern) -> List[Path]:
```
Comment (GirZ0n) [inline core/src/utils/file/file_utils.py]: Sorry, I used the wrong return type here. Please change it to `str`

File: jba/src/processing/tasktracker_content_collector.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        'tasks': [{
+            'name': 'example',
+            'description': 'description',
+            'id': 'main',
+            'files': files
+        }]
+    }
+
+
+def get_info_file(root: Path) -> Path:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `-> str`

File: jba/src/processing/tasktracker_content_collector.py [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    info_files = find_files_by_regex(root, INFO_FILE_REGEX)
+    return info_files[0]
+
+
+def get_files(course_root: Path, relative_path: Path, lesson: EduStructureNode) -> List[TaskTrackerFile]:
+    lesson_path = course_root / relative_path
+    info_file = get_info_file(lesson_path)
+    yaml_file_content = read_yaml_field_content(lesson_path / info_file, EduInfoFileField.TYPE.value)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `lesson_type = ...`

File: jba/src/processing/tasktracker_content_collector.py [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+    return itertools.chain.from_iterable(
+        [get_task_files(lesson_path / child.name,
+                        relative_path,
+                        yaml_file_content is not None and yaml_file_content == FRAMEWORK_TYPE)
+         for child in lesson.children])
+
+
+def get_task_files(root: Path, relative_path: Path, is_framework: bool) -> List[TaskTrackerFile]:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: For consistency, let's make this function accept the course root path (`root` -> `course_root`, now it accepts the task path) and the relative path to the task (`relative_path`, now it accepts the lesson relative path)

File: jba/src/processing/tasktracker_content_collector.py [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def get_files(course_root: Path, relative_path: Path, lesson: EduStructureNode) -> List[TaskTrackerFile]:
+    lesson_path = course_root / relative_path
+    info_file = get_info_file(lesson_path)
+    yaml_file_content = read_yaml_field_content(lesson_path / info_file, EduInfoFileField.TYPE.value)
+    return itertools.chain.from_iterable(
+        [get_task_files(lesson_path / child.name,
+                        relative_path,
+                        yaml_file_content is not None and yaml_file_content == FRAMEWORK_TYPE)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `None == FRAMEWORK_TYPE` returns false, so we can omit `yaml_file_content is not None`

File: jba/src/processing/tasktracker_content_collector.py [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+         for child in lesson.children])
+
+
+def get_task_files(root: Path, relative_path: Path, is_framework: bool) -> List[TaskTrackerFile]:
+    info_file = get_info_file(root)
+
+    files = read_yaml_field_content(root / info_file, EduInfoFileField.FILES.value)
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: Let's add the `default` argument to `read_yaml_field_content` that takes Any and defaults to None

Comment (GirZ0n) [reply jba/src/processing/tasktracker_content_collector.py]: So we could write `read_yaml_field_content(root / info_file, EduInfoFileField.FILES.value, [])`

File: jba/src/processing/tasktracker_content_collector.py [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+    if files is None:
+        files = []
+    files = list(filter(lambda file: file[EduInfoFileField.VISIBLE.value], files))
+
+    def get_filename(file_content: Dict) -> TaskTrackerFile:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py]: `get_filename` -> `convert_to_task_tracker_filer`

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+from core.src.utils.file.yaml_utils import parse_yaml
+from core.src.utils.subprocess_runner import run_in_subprocess
+from jba.src import MAIN_FOLDER
+
+from jba.tests.processing import PROCESSING_FOLDER
+
+import tempfile
+
+from deepdiff import DeepDiff  # noqa: SC200
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: Remove `# noqa: SC200` and add `deepdiff` to `spellcheck_whitelist.txt`

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+    with tempfile.TemporaryDirectory() as temp_directory:
+        prepare_course_directory = PROCESSING_FOLDER / 'collect_course_structure'
+        command = [sys.executable, (MAIN_FOLDER.parent / 'processing' / 'tasktracker_content_collector.py'),
+                   Path(temp_directory),
+                   prepare_course_directory / course_name]
+        stdout, stderr = run_in_subprocess(command)
+        assert stdout == ''
+        assert stderr == ''
+        assert len(os.listdir(temp_directory)) == 1
+        file = Path(temp_directory) / 'task_content_default.yaml'
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: `file` -> `actual_file`

File: jba/src/processing/tasktracker_content_collector.py [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+    lessons = course_structure.gather_leafs_of_type(EduStructureType.LESSON)
+    files = {
+        file
+        for path, lesson in lessons.items()
+        for file in get_files(course_root, Path(*path[1:]), lesson)
+    }
+    return get_data_template(list(map(lambda obj: obj.as_dict(), files)))
+
+
+def configure_parser(parser: argparse.ArgumentParser) -> None:
```
Comment (GirZ0n) [inline jba/src/processing/tasktracker_content_collector.py:101]: Let's swap `destination_path` and `course_sources_path`, because it's strange that you have to enter output first and then enter input

File: jba/tests/processing/test_tasktracker_content_collector.py (hunk 30 above)
Comment (GirZ0n) [reply jba/tests/processing/test_tasktracker_content_collector.py]: And use `CONTENT_FILE_NAME` here

File: jba/tests/processing/test_tasktracker_content_collector.py [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+    assert 'error: the following arguments are required' in stderr
+
+
+TEST_DATA = ['course_with_section', 'course_without_section', 'course_with_framework']
+
+
+@pytest.mark.parametrize('course_name', TEST_DATA)
+def test_course(course_name):
+    with tempfile.TemporaryDirectory() as temp_directory:
+        prepare_course_directory = PROCESSING_FOLDER / 'collect_course_structure'
```
Comment (GirZ0n) [inline jba/tests/processing/test_tasktracker_content_collector.py]: Use `COLLECT_COURSE_STRUCTURE_FOLDER`

**Review by GirZ0n — CHANGES_REQUESTED** [review]

File: core/src/utils/file/file_utils.py (hunk 22 above)
Comment (mikrise2) [reply core/src/utils/file/file_utils.py]: I think the return type is correct

File: jba/src/processing/tasktracker_content_collector.py (hunk 23 above)
Comment (mikrise2) [reply jba/src/processing/tasktracker_content_collector.py]: Here return type is also correct

File: jba/src/processing/tasktracker_content_collector.py (hunk 25 above)
Comment (mikrise2) [reply jba/src/processing/tasktracker_content_collector.py]: I suggest simply renaming 'root' to 'task_root.' I believe there is no point in passing the project root since it is not used anywhere. What we need is the path to the task file and the relative path from the root to the files.

File: core/src/utils/file/file_utils.py (hunk 22 above)
Comment (GirZ0n) [reply core/src/utils/file/file_utils.py]: I mean we should change it to `List[str]`, because `os.listdir` returns a list of strings.

image

**Review by GirZ0n — APPROVED** [review]
Thank you! Now everything looks good!
